subsys/power/power.c

```diff
@@ -81,6 +87,10 @@ void sys_pm_force_power_state(enum power_states state)
 		 "Invalid power state %d!", state);
 
 	forced_pm_state = state;
+
+#ifdef CONFIG_SYSTEM_PM_WORKQUEUE
```

Comment:
@vanti I believe there is the case triggering system suspend in interrupt context where maybe _sys_suspend isn't suitable to execute. And also the latency of idle thread is from waiting other threads suspend, and as usual the workqueue will be in high priority, so maybe the latency of workqueue will be one context switch.

Reply:
The context of the caller of sys_pm_force_power_state() will be running when the function is invoked, so there won't be any waiting for the other threads to suspend, regardless of the caller's priority. And given you are disabling interrupts prior to calling _sys_suspend(), there is no latency from servicing interrupts either. Thus a direct call to _sys_suspend() would be the fastest way to get to suspend, with zero context switch.

For the users who just want to suspend the system from a thread context, bringing in workqueues seems like a significant penalty. I'd suggest to either make the workqueue an optional alternative to a direct call to _sys_suspend(), or (better yet) remove it and simply let the application create its own thread to call sys_pm_force_power_state(), that the interrupt context would make ready if suspend is desired. It would just be standard "bottom-half" interrupt handling, and the application can prioritize the thread as highly as necessary.